nasa/ow_simulator
# Oceanwater 996 material integrator

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-812](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-812) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-996](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-996) |

## Sibling PRs
[ow_europa #26](https://github.com/nasa/ow_europa/pull/26)

## Summary of Changes
* Correlating dug positions with the material grid, the bread-and-butter of the MaterialDistributionPlugin, is completed.
* Visualization elements added to the default Rviz configuration (unticked by default):
  * Stereo cam point cloud
  * Multi-material grid
  * Dug points cloud

## Caveats
The Rviz visualizations currently only work in atacama_y1a, because a transform had to be hard coded. This is due to a bug in the TF tree described by [ow-1194](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1194). There is probably a smarter work around for this, but I would rather keep it simple with the hope that we can get OW-1194 fixed before this epic branch is complete. 

The resulting dug point cloud varies from run to run, and contains large voids that should be filled with points. This may be a bug of its own, but I suspect it is at least in part related to [OW-887](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-887), number of regolith particles varying, and I think it comes back to both of these nodes missing modification messages because they just come in too fast. I plan to improve on this later in the Epic branch.  

## Test
1. Launch atacama_y1a with Rviz 
2. In Rviz, tick "material_grid" to on in the "Displays" menu. Your framerate in rviz will drop precipitous, this is normal since Rviz is rendering over 6 million cubes. You should see
![Screenshot_20230818_120014](https://github.com/nasa/ow_simulator/assets/17039973/ae628136-85a9-4a60-9d32-d95ec96b7e0c)
The 3 colors represent the 3 different materials that are loaded into the plugin and given bright colors for demonstration purposes. 
3. Untick "material_grid" and tick "dug_points". 
4. Call a default TaskGrind `rosrun ow_lander task_grind.py`. By the end of the action you will see. 
![Screenshot_20230818_120155](https://github.com/nasa/ow_simulator/assets/17039973/eee47676-e3ab-4af6-b3fd-a66a7033a340)
In the boundaries between two/three different colors you should see the points take on a blended color of all materials involved in that frame of that dig e.g. orange or purple. This demonstrates how properties of dug materials will be blended to generate a bulk material property.
5. Call a default TaskScoopLinear `rosrun ow_lander task_scoop_linear.py`
![Screenshot_20230818_120312](https://github.com/nasa/ow_simulator/assets/17039973/c354f7b3-31df-4748-807c-97b7dec53e1e)
